extension/source/Controllers/Keyring/KeyringController.ts

```diff
@@ -68,5 +60,49 @@ export default class KeyringController
     }
   }
 
-  // TODO: add methods to sign transactions / messages
+  async signTransactions(address: string, tx: Operation) {
+    const signer = await this._getBLSSinger();
+    const privKey = this._getPrivateKeyFor(address);
+
+    return signer.sign(tx, privKey);
+  }
+
+  async _createAccountAndUpdate(privateKey: string): Promise<string> {
+    const address = await this._getContractWalletAddress(privateKey);
+
+    if (address) {
+      this.update({
+        wallets: [
+          ...this.state.wallets,
+          {
+            privateKey,
+            address: address,
+          },
+        ],
+      });
+    }
+
+    return address;
+  }
+
+  private _getPrivateKeyFor(address: string): string {
+    const keyPair = this.state.wallets.find((x) => x.address === address);
+    if (!keyPair) throw new Error('key does not exist');
+    return keyPair.privateKey;
+  }
+
+  private _getContractWalletAddress(privateKey: string) {
+    const provider = new ethers.providers.JsonRpcProvider(PROVIDER_URL);
+    return BlsWalletWrapper.Address(
+      privateKey,
+      NETWORK_CONFIG.addresses.verificationGateway,
+      provider,
+    );
+  }
+
+  private async _getBLSSinger() {
+    return initBlsWalletSigner({
+      chainId: Number(this.state.chainId),
+    });
+  }
```

Comment:
Instead of initing and handling signing via the BLS signer directly, you may just want to use the `BlsWalletWrapper` abstraction instead and add it into the state. I am unsure if the state needs to have only primitively typed data in it, if so you may be able to use another controller/construct to manage `BlsWalletWrapper` wallet objects.

You then would only need to call `BlsWalletWrapper.connect` once and then store the wallet object it returns and use that for signing/address operations. See https://github.com/jzaki/bls-wallet/tree/main/contracts/clients#blswalletwrapper for details.

Reply:
semi-fixed in f30cd85

Yeah I agree putting it in state and then just reusing the same object as the provider doesn't change.  
As of now I am creating a new object with a new provider for simplicity sakes. Would like to take @chaitanyapotti's opinion on how to implement it

Reply:
semi-fixed in f30cd85

Yeah I agree. As the prev comment mentions, should use the provider coming from the state. Couldn't because of incompatibility. Would be nice to get @chaitanyapotti's opinion on how to pass the provider to each controller.

Also @jacque006, the wrapper currently has a private constructor and static connect and address methods. It wont be possible to create an object (with only provider and verification address), save it in state and just use different privKeys to do different ops. WDYT about adding these to the client library itself?

Reply:
`connect` is essentially the async constructor for `BLSWalletWrapper` as it returns an initialized `BLSWalletWrapper` object. You would need to `connect` for each private key and store the wrapper object for each wallet.

I think rolling with the static methods for now is fine per our offline conversation. But it may be easier in the future to have a common place the `BLSWalletWrapper` object(s) can be accessed from.